bundles/persistence/org.eclipse.smarthome.persistence.mapdb/src/main/java/org/eclipse/smarthome/persistence/mapdb/internal/MapDbPersistenceService.java

```diff
@@ -0,0 +1,176 @@
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.eclipse.smarthome.core.items.Item;
+import org.eclipse.smarthome.core.persistence.FilterCriteria;
+import org.eclipse.smarthome.core.persistence.HistoricItem;
+import org.eclipse.smarthome.core.persistence.PersistenceItemInfo;
+import org.eclipse.smarthome.core.persistence.PersistenceService;
+import org.eclipse.smarthome.core.persistence.QueryablePersistenceService;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.mapdb.DB;
+import org.mapdb.DBMaker;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * This is the implementation of the MapDB {@link PersistenceService}. To learn
+ * more about MapDB please visit their <a
+ * href="http://www.mapdb.org/">website</a>.
+ *
+ * @author Jens Viebig - Initial contribution
+ * @author Martin Kühl - Port to Eclipse SmartHome
+ */
+@Component(service = PersistenceService.class)
+public class MapDbPersistenceService implements QueryablePersistenceService {
+
+    private static final @NonNull String SERVICE_NAME = "mapdb";
+
+    private static final String DB_FOLDER_NAME = ConfigConstants.getUserDataFolder() + File.separator + "persistence" + File.separator + "mapdb";
+
+    private static final String DB_FILE_NAME = "storage.mapdb";
+
+    private static final Logger logger = LoggerFactory.getLogger(MapDbPersistenceService.class);
+
+    @NonNullByDefault({})
+    private ExecutorService threadPool;
+
+    /** holds the local instance of the MapDB database */
+    private DB db;
+    private Map<String, String> map;
+
+    private transient Gson mapper = new GsonBuilder()
+            .registerTypeAdapter(State.class, new StateTypeAdapter())
+            .create();
+
+    public void activate() {
+        logger.debug("MapDB persistence service is being activated");
+
+        threadPool = ThreadPoolManager.getPool(getClass().getSimpleName());
+
+        File folder = new File(DB_FOLDER_NAME);
+        if (!folder.exists()) {
+            if (!folder.mkdirs()) {
+              logger.error("Failed to create one or more directories in the path '{}'", DB_FOLDER_NAME);
+              logger.error("MapDB persistence service activation has failed.");
+              return;
+            }
+        }
+
+        File dbFile = new File(DB_FOLDER_NAME, DB_FILE_NAME);
+        db = DBMaker.newFileDB(dbFile).closeOnJvmShutdown().make();
+        map = db.createTreeMap("itemStore").makeOrGet();
+        logger.debug("MapDB persistence service is now activated");
+    }
+
+    public void deactivate() {
+        logger.debug("MapDB persistence service deactivated");
+        if (db != null) {
+            db.close();
+        }
+        threadPool.shutdown();
+    }
+
+    @Override
+    public @NonNull String getId() {
```

Comment:
please use `@NonNullByDefault` on class level and then use `@Nullabel` accordingly.